### PR TITLE
better error help when schema is in the wrong format

### LINKED ISSUE #779: Better error message when your schema is the wrong format
### Category

Error message improvements

### Describe the feature you'd like to request

Better error message when you pass a JSON schema where human is expected, or vice versa.  Especially in the CLI, given #750; but also in the Rust public API.  We might particularly notice in the CLI when the filename ends with `.json`, but `--schema-format json` is not passed.

### Describe alternatives you've considered

Fully auto-detect whether you're using human or natural schema format, at least in the CLI

### Additional context

_No response_

### Is this something that you'd be interested in working on?

- [ ] 👋 I may be able to implement this feature request
- [ ] ⚠️ This feature might incur a breaking change

### PR DESCRIPTION
**Pull request by cdisselkoen**
## Description of changes

Improves the error message when we think (based on simple heuristic) you might have passed a JSON schema where a human-format one is expected, or vice versa.

Breaking changes to error types.

## Issue #, if available

Fixes #779

## Checklist for requesting a review

The change in this PR is (choose one, and delete the other options):

- [x] A breaking change requiring a major version bump to `cedar-policy` (e.g., changes to the signature of an existing API).

I confirm that this PR (choose one, and delete the other options):

- [x] Does not update the CHANGELOG because my change does not significantly impact released code.

I confirm that [`cedar-spec`](https://github.com/cedar-policy/cedar-spec) (choose one, and delete the other options):

- [x] Requires updates, and I have made / will make these updates myself. (Please include in your description a timeline or link to the relevant PR in `cedar-spec`, and how you have tested that your updates are correct.)

### CODE REVIEW

**Review by shaobo-he-aws — APPROVED** [review]

**Review by john-h-kastner-aws — COMMENTED** [review]
Should add a couple test cases

File: cedar-policy-validator/src/err.rs [hunk 1]
```diff
@@ -24,25 +24,100 @@ use itertools::Itertools;
...
+            Some('{') => true, // yes, this looks like it was intended to be a JSON schema
+            Some(_) => true, // any character other than '{', not likely it was intended to be a JSON schema
+        };
+        Self {
+            errs,
+            suspect_json_format,
+        }
+    }
+
+    #[cfg_attr(not(test), allow(dead_code))]
```
[the remark this answers is not in the document]
Comment (cdisselkoen) [reply cedar-policy-validator/src/err.rs:109]: Lol. Thanks.

File: cedar-policy-validator/src/err.rs [hunk 2]
```diff
@@ -24,25 +24,100 @@ use itertools::Itertools;
...
+impl HumanSyntaxParseError {
+    /// `errs`: the `human_schema::parser::HumanSyntaxParseErrors` that were thrown
+    ///
+    /// `src`: the human-syntax text that we were trying to parse
+    pub(crate) fn new(errs: human_schema::parser::HumanSyntaxParseErrors, src: &str) -> Self {
+        // let's see what the first non-whitespace character is
+        let suspect_json_format = match src.trim_start().chars().next() {
+            None => false, // schema is empty or only whitespace; the problem is unlikely to be JSON vs human format
+            Some('{') => true, // yes, this looks like it was intended to be a JSON schema
+            Some(_) => true, // any character other than '{', not likely it was intended to be a JSON schema
```
[the remark this answers is not in the document]
Comment (cdisselkoen) [reply cedar-policy-validator/src/err.rs]: Fixed.

File: cedar-policy-validator/src/err.rs [hunk 3]
```diff
@@ -172,3 +247,50 @@ pub enum UnsupportedFeature {
...
+    err: serde_json::Error,
+    /// Did the schema look like it was intended to be human format instead of
+    /// JSON?
+    suspect_human_format: bool,
+}
+
+impl Diagnostic for JsonDeserializationError {
+    fn help<'a>(&'a self) -> Option<Box<dyn std::fmt::Display + 'a>> {
+        if self.suspect_human_format {
+            Some(Box::new("this API was expecting a schema in the JSON format; did you mean to use `from_natural_str()` or similar, which expects the Cedar schema format?"))
```
[the remark this answers is not in the document]
Comment (cdisselkoen) [reply cedar-policy-validator/src/err.rs]: Probably wise. Fixed.

File: cedar-policy/src/tests.rs [hunk 4]
```diff
@@ -3294,6 +3295,38 @@ mod error_source_tests {
...
+    use crate::Schema;
+    use cool_asserts::assert_matches;
+    use miette::Diagnostic;
+
+    #[test]
+    fn issue_779() {
+        let json = r#"{ "" : { "actions": { "view": {} }, "entityTypes": { invalid } }}"#;
+        let human = r#"namespace Foo { entity User; action View; invalid }"#;
+
+        assert_matches!(Schema::from_json_str(human), Err(e) => {
```
Comment (john-h-kastner-aws) [inline cedar-policy/src/tests.rs:3308]: With #803 we can use `test_utils::expect_err` in this crate

**Review by john-h-kastner-aws — APPROVED** [review]

**Pull request merged**